ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -688,38 +741,58 @@ genChainUpdates securityParam n =
 
     genChainUpdate = do
       ChainUpdateState { cusCurrentChain = chain } <- get
-      frequency'
-        [ (3, genAddBlock)
-        , (if Chain.null chain then 0 else 1, genSwitchFork)
-        ]
-
-    genForkNo = frequency
-      [ (1, return 0)
-      , (1, choose (1, 2))
-      ]
+      let genValid =
+            frequency'
+              [ (3, genAddBlock Valid)
+              , ( if Chain.null chain then 0 else 1
+                , genSwitchFork (choose (1, k))
+                )
+              ]
+      frequency' $
+        (5, replicateM_ 2 genValid) :
+        [ (1, genInvalidBlock) | updateBehavior == TentativeChainBehavior ]
+
+    genBlockToAdd validity = do
+        ChainUpdateState { cusCurrentChain = chain } <- get
+        block <- lift $ case Chain.head chain of
+          Nothing      -> setValidity . firstBlock <$> genForkNo
+          Just curHead -> do
+            forkNo <- case validity of
+              Valid   ->  genForkNo
+              Invalid -> pure 3
+            return
+              . modifyFork (const forkNo)
+              . setValidity
+              $ successorBlock curHead
+        modify $ setChain (Chain.addBlock block chain)
+        return block
+      where
+        setValidity b = b { tbValid = validity }
+        genForkNo = case validity of
+          Valid -> frequency
+            [ (1, return 0)
+            , (1, choose (1, 2))
+            ]
+          -- Blocks with equal hashes have to have equal validity, so we reserve
+          -- a specific ForkNo for invalid blocks to ensure this.
+          Invalid -> pure 3
 
-    genBlockToAdd = do
-      ChainUpdateState { cusCurrentChain = chain } <- get
-      block <- lift $ case Chain.head chain of
-        Nothing      -> firstBlock <$> genForkNo
-        Just curHead -> do
-          forkNo <- genForkNo
-          return $ modifyFork (const forkNo) (successorBlock curHead)
-      modify $ setChain (Chain.addBlock block chain)
-      return block
-
-    genAddBlock = do
-      block <- genBlockToAdd
+    genAddBlock isValid = do
+      block <- genBlockToAdd isValid
       modify $ addUpdate (AddBlock block)
 
-    genSwitchFork  = do
+    genSwitchFork genRollBackBlocks = do
       ChainUpdateState { cusCurrentChain = chain } <- get
-      rollBackBlocks <- lift $ choose (1, k)
+      rollBackBlocks <- lift genRollBackBlocks
       let chain' = Chain.drop rollBackBlocks chain
       modify $ setChain chain'
-      blocks <- replicateM rollBackBlocks genBlockToAdd
+      blocks <- replicateM rollBackBlocks (genBlockToAdd Valid)
```

Comment:
Do we _never_ want to switch from one invalid fork to another invalid fork?

I imagine we discussed this on our phone call last week, but I do not remember (got spiked 😞). So, if we never do: we should comment here explaining why :)

Reply:
I don't think we discussed this, but it it indeed possible honest behavior to switch to a *better* invalid fork (i.e. we receive a better trap header). I will add this possibility later today.

Reply:
We discussed this on a call and decided to include this expressivity boost in https://input-output.atlassian.net/browse/CAD-4314

So I'll resolve this comment thread.